### DISCUSSION TITLE
Turbo Broadcastables & Helpers

### DISCUSSION THREAD

**Discussion by gabriel-curtino** [yippee-fun/phlex#806] [Help] [answered]
Hi!

Before anything, thanks for your work! I'm trying `phlex-rails`  for first time and it looks really interesting. 

I'm using Phlex in a rails app and currently facing an issue when trying to broadcast from the model (https://www.rubydoc.info/gems/turbo-rails/Turbo/Broadcastable)

Broadcasting the `html` works, but it fails when the component use helpers: 

```ruby
broadcast_append_to(channel_id, html: Components::ChatMessage.new(self).call, target: "thread_#{thread.id}_messages")

# undefined method `class_names' for nil (NoMethodError)
```

I'm using a workaround to render the component as if it would be in a controller context:

```ruby
def render_component(component)
  ApplicationController.render(component, layout: false)
end

broadcast_append_to(channel_id, html: render_component(Components::ChatMessage.new(self)), target: "thread_#{thread.id}_messages")
```

Is there a proper way to do it? Couldn't find an answer in the docs or the discussions.

EDIT: using the v2 `main` branch

**Comment by joeldrapper** [comment]
I haven't tried it, but I wonder if you could use `render_to_string`. https://api.rubyonrails.org/classes/AbstractController/Rendering.html#method-i-render_to_string

**Comment by gabriel-curtino** [comment]
Testing in the console:

```ruby 
ApplicationController.render_to_string(Components::ChatMessage.new(Message.first))

# nil
```

```ruby 
ApplicationController.render_to_string(Components::ChatMessage.new(Message.first).call)

# `view_template': undefined method `class_names' for nil (NoMethodError)
```

The following alternatives work:

```ruby
Components::ChatMessage.new(Message.first).render_in(ActionController::Base.new.view_context)
```

```ruby
ApplicationController.new.view_context.render(Components::ChatMessage.new(Message.first))
```

I got the ideas from ViewComponents here: https://github.com/ViewComponent/view_component/issues/1106

**Comment by fagnerpereira** [comment, marked as answer]
try to use the `:renderable` following the example provided by the turbo rails doc:

```ruby
broadcast_replace_to(user, :message, target: "message", renderable: MessageComponent.new)
```

PS: I just implemented here and it worked perfectly well.

**Reply by joeldrapper** [reply]
Oh this is great. We should get this in the Phlex docs. I didn't realise you could do this.

**Reply by gabriel-curtino** [reply]
I missed that one! Will try it, thanks!

**Reply by gabriel-curtino** [reply]
It does work, except when you defer the broadcast: 

```ruby
>> broadcast_append_later_to(channel_id, :message, renderable: Components::ChatMessage.new(self), target: "conversation_#{conversation_id}_messages")

#<ActiveJob::SerializationError: Unsupported argument type: Components::ChatMessage> || /ruby/3.3.4/lib/ruby/gems/3.3.0/gems/activejob-8.0.0.beta1/lib/active_job/serializers.rb:34:in `serialize'
```

It seems it doesn't support these cases.

**Reply by joeldrapper** [reply]
It would work if the component was serialisable for ActiveJob. Let’s see if we can get that working.

**Reply by gabriel-curtino** [reply]
That's correct, but does it worth to serialize the whole object? I guess it might only if the component does heavy operations and the rendering takes a while, otherwise seems that pass the html to the job might be better.  

Anyway, it will be nice to have that possibility and explore how convenient it is.

**Reply by fagnerpereira** [reply]
@joeldrapper there is a pr opened to support that, but it seens they are not advancing with that https://github.com/hotwired/turbo-rails/pull/501

**Reply by gabriel-curtino** [reply]
> @joeldrapper there is a pr opened to support that, but it seens they are not advancing with that [hotwired/turbo-rails#501](https://github.com/hotwired/turbo-rails/pull/501)

That PR renders the element synchronously and then pass the html to the job. The only benefit is that saves you from doing it yourself, but can be misleading: you might think the rendering will be done asynchronously...

**Reply by joeldrapper** [reply]
I think there's a way we could allow the component itself to be serialized.
